sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -839,6 +839,17 @@ object SQLConf {
     .checkValues(HiveCaseSensitiveInferenceMode.values.map(_.toString))
     .createWithDefault(HiveCaseSensitiveInferenceMode.NEVER_INFER.toString)
 
+  val HIVE_SUPPORTED_SCHEMES_TO_USE_NONBLOBSTORE =
+    buildConf("spark.sql.hive.supportedSchemesToUseNonBlobstore")
+      .doc("Comma-separated list of supported blobstore schemes (e.g. 's3,s3a'). " +
+        "If any blobstore schemes are specified, this feature is enabled. " +
+        "When writing data out to a Hive table, " +
+        "Spark writes the data first into non blobstore storage, and then moves it to blobstore. " +
+        "By default, this option is set to empty. It means this feature is disabled.")
+      .version("3.1.0")
+      .stringConf
+      .createWithDefault("")
```

Comment:
The supported schemes are determined by Hive? If so, is it meaningful to config the supported schemes here?

Reply:
Users can specify any blob storage schema like following. If copy operation is expensive in the storage system, this option will be effective.
- Amazon S3: `s3`, `s3a`, `s3n`
- Azure Blob Storage: `wasb`, `wasbs`
- Google Cloud Storage: `gs`
- Databricks: `dbfs`
- OpenStack: `swift`

Since any schemes are possible to be used, I believe we cannot define specific supported schemes here. That's why I just listed samples in SQLConf.scala.

Reply:
Note: I am not 100% sure whether all these blob storage systems have similar characteristics and not sure if this option is effective. At least, this option is effective for Amazon S3.

Reply:
If users can specify any blob storage scheme, I'm not sure if it is meaningful to have a list config of schemes. It sounds like we just need a boolean config to enable/disable this feature.

Reply:
Ah, nice. +1 on the @viirya idea.

Reply:
If all the blob storage services have the same characteristics, then it will be possible. However, actually we cannot guarantee it.
I believe that it is better to allow users to enable/disable this feature per scheme.

Reply:
Ah, if we use a boolean config for this, we cannot apply this optimization only at a specific schema... I see.

Reply:
Okay, it sounds making sense.